modules/memfault-firmware-sdk/memfault_location_metrics.c

```diff
@@ -0,0 +1,209 @@
+{
+	prv_location_method_results_record(method, fallback->cause, &fallback->details, NULL);
+
+	/* Log the results */
+	char method_str[16];
+
+	snprintk(method_str, sizeof(method_str), "%s", prv_location_method_to_string(method));
+	LOG_DBG("%s method attempted", method_str);
+
+	char fallback_method_str[16];
+
+	snprintk(fallback_method_str, sizeof(fallback_method_str), "%s",
+		 prv_location_method_to_string(fallback->next_method));
+	LOG_DBG("Falling back to %s method", fallback_method_str);
+}
+
+/**
+ * @brief Start a location session.
+ *
+ */
+static void prv_location_metrics_session_start(void)
+{
+	if (s_session_in_progress) {
+		return;
+	}
+
+	s_session_in_progress = true;
+	LOG_DBG("Starting location session");
+	MEMFAULT_METRICS_SESSION_START(ncs_loc);
+	MEMFAULT_METRIC_ADD(ncs_loc_search_request_count, 1);
+}
+
+/**
+ * @brief Stop a location fix session.
+ *
+ * @param event_data pointer to location event data
+ */
+static void prv_location_metrics_session_stop(const struct location_event_data *event_data)
+{
+	if (!s_session_in_progress) {
+		return;
+	}
+
+	switch (event_data->id) {
+	case LOCATION_EVT_LOCATION:
+		LOG_DBG("Stopping location session, fix data acquired");
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_search_time_ms,
+					     event_data->location.details.elapsed_time_method);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_search_success, 1);
+
+		prv_location_method_results_record(event_data->method, event_data->id,
+						   &event_data->location.details,
+						   &event_data->location.accuracy);
+
+		break;
+	case LOCATION_EVT_TIMEOUT:
+		/* Intentional fall through */
+	case LOCATION_EVT_ERROR:
+		/* Intentional fall through */
+	case LOCATION_EVT_RESULT_UNKNOWN:
+		/* Intentional fall through */
+	default:
+		LOG_DBG("Stopping location session, no fix found, id=%d", event_data->id);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_search_failure, 1);
+		prv_location_method_results_record(event_data->method, event_data->id,
+						   &event_data->error.details, NULL);
+		break;
+	}
+
+	MEMFAULT_METRICS_SESSION_END(ncs_loc);
+	s_session_in_progress = false;
+}
+
+/**
+ * @brief Handle location events
+ *
+ * @param event_data pointer to location event data
+ */
+static void prv_location_event_handler(const struct location_event_data *event_data)
+{
+	if (event_data->id == LOCATION_EVT_STARTED) {
+		prv_location_metrics_session_start();
+	} else if (event_data->id == LOCATION_EVT_FALLBACK) {
+		prv_location_method_fallback_record(event_data->method, &event_data->fallback);
+	} else if (event_data->id == LOCATION_EVT_LOCATION ||
+		   event_data->id == LOCATION_EVT_TIMEOUT || event_data->id == LOCATION_EVT_ERROR ||
+		   event_data->id == LOCATION_EVT_RESULT_UNKNOWN) {
+		prv_location_metrics_session_stop(event_data);
+	}
+}
+
+void memfault_location_metrics_init(void)
+{
+	int result = location_handler_register(prv_location_event_handler);
+
+	if (result != 0) {
+		LOG_DBG("Error registering location handler, err %d", result);
```

Comment:
I guess it would be good to raise the attention of this failure by making it error log. It's very abnormal to get a failure here.

```suggestion
		LOG_ERR("Error registering location handler, err %d", result);
```